keps/sig-api-machinery/0030-storage-migration.md

```diff
@@ -256,28 +249,28 @@ Before upgrading or downgrading the cluster, the cluster administrator must run
 migrations have completed. Otherwise the apiserver can crash, because it cannot
 interpret the serialized data in etcd. To mitigate, the cluster administrator
 can rollback the apiserver to the old version, and wait for the migration to
-complete. Even if the apiserver does not crash after upgrading or downgrading,
-the `migration` objects are not accurate anymore, because the default storage
-versions might have changed after upgrading or downgrading, but no one
-increments the `migration.spec.generation`.  Administrator needs to re-run the
-`kubectl run migrate --image=migrator-initializer --restart=OnFailure` command
-to recover.
-
-TODO: it is safe to rollback an apiserver to the previous configuration without
-waiting for the migration to complete. It is only unsafe to roll-forward or
-rollback twice. We need to design how to record the previous configuration.
+complete. 
 
-## Graduation Criteria
+With the newly introduced [storageStates API][], the cluster administrator can
+fast upgrade/downgrade as long as the new apiserver binaries understand all
+versions recorded in storageState.status.persistedStorageVersionHashes.
```

Comment:
There's no practical way to know that today, right?

Reply:
No, the apiserver binary does not expose this information anywhere.

Reply:
api server exposes all versions it understands in its discovery doc, right?

Reply:
No, apiservers can understand more versions. A concrete example is that in 1.15, the workload API in extensions/v1beta1 will be removed from the discovery doc, but the apiserver still understands them.